Framework/Algorithms/src/MultipleScatteringCorrection.cpp

```diff
@@ -236,111 +515,238 @@ void MultipleScatteringCorrection::exec() {
+    } else {
+      msg_debug << "Sample element " << idx - numVolumeElementsContainer << "\n";
+    }
+    msg_debug << "LS1_container=" << LS1sContainer[idx] << ", LS1_sample=" << LS1sSample[idx] << "\n";
+    g_log.notice(msg_debug.str());
+#endif
+  }
+}
 
-  // Get the beam direction
-  m_beamDirection = m_inputWS->getInstrument()->getBeamDirection();
+/**
+ * @brief calculate L12 for single component case
+ *
+ * @param distGraber
+ * @param L12s
+ * @param shape
+ */
+void MultipleScatteringCorrection::calculateL12s(const MultipleScatteringCorrectionDistGraber &distGraber, //
+                                                 std::vector<double> &L12s,                                //
+                                                 const Geometry::IObject &shape) {
+  const int64_t numVolumeElements = distGraber.m_numVolumeElements;
+  // L12 is independent from the detector, therefore can be cached outside
+  // - L12 is a upper off-diagonal matrix
+  // NOTE: if the sample size/volume is too large, we might need to use openMP
+  //       to parallelize the calculation
 
-  // Get the total number of wavelength points, default to use all if not specified
-  m_num_lambda = isDefault("NumberOfWavelengthPoints") ? static_cast<int64_t>(m_inputWS->blocksize())
-                                                       : getProperty("NumberOfWavelengthPoints");
-  // -- while we're here, compute the step in bin number between two adjacent points
-  const auto specSize = static_cast<int64_t>(m_inputWS->blocksize());
-  m_xStep = std::max(int64_t(1), specSize / m_num_lambda); // Bin step between points to calculate
-  // -- notify the user of the bin step
-  std::ostringstream msg;
-  msg << "Numerical integration performed every " << m_xStep << " wavelength points";
-  g_log.information() << msg.str();
+  PARALLEL_FOR_IF(Kernel::threadSafe(*m_inputWS))
+  for (int64_t indexTo = 0; indexTo < numVolumeElements; ++indexTo) {
+    PARALLEL_START_INTERUPT_REGION
 
-  // Get the element size
-  m_elementSize = getProperty("ElementSize"); // in mm
-  m_elementSize = m_elementSize * 1e-3;       // convert to m
+    const auto posTo = distGraber.m_elementPositions[indexTo];
+    Track track(posTo, V3D{0, 0, 1}); // take object creation out of the loop
+    for (int64_t indexFrom = indexTo + 1; indexFrom < numVolumeElements; ++indexFrom) {
+      // where in the final result to update, e.g.
+      // x 1 2 3
+      // x x 4 5
+      // x x x 6
+      // x x x x
+      const int64_t idx = calcLinearIdxFromUpperTriangular(numVolumeElements, indexTo, indexFrom);
+
+      const auto posFrom = distGraber.m_elementPositions[indexFrom];
+      const V3D unitVector = getDirection(posFrom, posTo);
+
+      // reset information in the Track and calculate distance
+      track.reset(posFrom, unitVector);
+      track.clearIntersectionResults();
+      const auto distance1 = getDistanceInsideObject(shape, track);
```

Comment:
Maybe rename this to `rayLengthOne1` and `rayLengthOne2` since the distance is the difference between them. Adding a comment explaining why wouldn't hurt.

Reply:
Two local vars are renamed, and the doc is added as suggested.